PC/launcher2.c

```diff
@@ -871,6 +871,77 @@ _findCommand(SearchInfo *search, const wchar_t *command, int commandLength)
 }
 
 
+int
+_useShebangAsExecutable(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+{
+    wchar_t buffer[MAXLEN];
+    wchar_t script[MAXLEN];
+    wchar_t command[MAXLEN];
+
+    int commandLength = 0;
+    int inQuote = 0;
+    int afterSlash = 0;
+
+    if (!shebang || !shebangLength) {
+        return 0;
+    }
+
+    wchar_t *pC = command;
+    for (int i = 0; i < shebangLength; ++i) {
+        wchar_t c = shebang[i];
+        if ((isspace(c) || c == L'\r' || c == L'\n') && !inQuote) {
+            commandLength = i;
+            break;
+        } else if (c == L'"') {
+            if (!afterSlash) {
+                // non-escaped quote. either way, we don't add it to the path
+                inQuote = !inQuote;
+            }
+            afterSlash = 0;
+        } else if (c == L'/' || c == L'\\') {
+            if (afterSlash) {
+                // escaped slash
+                *pC++ = L'\\';
+                afterSlash = 0;
+            } else {
+                afterSlash = 1;
+            }
+        } else {
+            if (afterSlash) {
+                *pC++ = L'\\';
+                afterSlash = 0;
+            }
+            *pC++ = c;
+        }
```

Comment:
I think you're making this more complicated than it has to be. There is no escaping of double quotes or path separators in the filepath part of a shebang. They're illegal characters in a filename. Also, you can get the system to parse the shebang into the command and parameters via [`CommandLineToArgvW()`](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw).

Reply:
We need to allow quotes on Windows for this to be at all useful (in scenarios that aren't covered by `/usr/bin/env`). Again, it's not portable from POSIX at all, but there's not really any way that it can be.

CommandLineToArgvW is probably just as complicated, since what we really need is the index of the first argument, and it doesn't give that to us. Skipping to the end of the first quoted argument is easier than trying to re-join already split arguments correctly.

Reply:
Adding double quotes for commands that contain spaces is handled by `calculateCommandLine()`. My comment was only about parsing `executablePath` from the command line. The code is overly complicated and mistaken with the `afterSlash` handling, which should be removed. A double quote always toggles quote mode when parsing `argv[0]` from the command line. For example:

```pycon
>>> subprocess.call(r'"foo\" -iq"', executable=sys.executable)
>>> import sys
>>> sys.orig_argv
['foo\\', '-iq']
```
If the second double quote didn't toggle quote mode, then the parsed `argv[0]` would have been `'foo\\ -iq'`.

On closer inspection, it turns out that `CommandLineToArgvW()` has problems. The C runtime developers should be asked to update it to better match how the CRT parses the command line. Its handling of quote mode for parsing `argv[0]` is too simplistic. If a command begins with a double quote, then it consumes up to the next double quote or the end of the command line. If a command doesn't begin with a double quote, then it consumes up to the first space or the end of the command line, and double quotes are left in `argv[0]` as literal characters. For example, the following results are wrong relative to what the C runtime does:

```pycon
>>> win32api.CommandLineToArgv(r'"foo"bar')
['foo', 'bar']
>>> win32api.CommandLineToArgv(r'spam""" "foo bar"')
['spam"""', 'foo bar']
```

Reply:
> The first argument (argv[0]) is treated specially...

Where are you reading this from? I believe it, I just haven't seen it written down like this before.

Even so, it's only going to simplify the code slightly. Unquoting and then requoting each argument isn't something I want to be doing, and `calculateCommandLine` doesn't do it - it only handles `argv[0]`.

Reply:
> Where are you reading this from? I believe it, I just haven't seen it written down like this before.

The documentation that I quoted is here: https://learn.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments?view=msvc-170. 

The source code is in `parse_command_line()` in "ucrt/startup/argv_parsing.cpp". Here's the loop that parses out `argv[0]`:

```c
    /* first scan the program name, copy it, and count the bytes */
    Character* p = cmdstart;
    if (argv)
        *argv++ = args;

    // A quoted program name is handled here. The handling is much
    // simpler than for other arguments. Basically, whatever lies
    // between the leading double-quote and next one, or a terminal null
    // character is simply accepted. Fancier handling is not required
    // because the program name must be a legal NTFS/HPFS file name.
    // Note that the double-quote characters are not copied, nor do they
    // contribute to character_count.
    bool in_quotes = false;
    do
    {
        if (*p == '"')
        {
            in_quotes = !in_quotes;
            c = *p++;
            continue;
        }

        ++*character_count;
        if (args)
            *args++ = *p;

        c = *p++;

        if (should_copy_another_character(c))
        {
            ++*character_count;
            if (args)
                *args++ = *p; // Copy 2nd byte too
            ++p; // skip over trail byte
        }
    }
    while (c != '\0' && (in_quotes || (c != ' ' && c != '\t')));
```

`should_copy_another_character()` is always false for the wide-character version since none of the checked characters is a UTF-16 surrogate pair. Note that it doesn't use `isspace()`, so LF, VT, FF, CR, and no-break space (160) don't count as white space.

Reply:
As mentioned in the comment, they assume it's already a valid path (though they shouldn't...). In our case, I think `isspace()` is appropriate for breaking the shebang line, even if technically you could force the driver to create the path.

I've already simplified the referenced code here, and I don't think any of these helper functions will be further simplification. It works, so I'll merge, but if there's a case that breaks things that I missed, please file another bug pointing it out and I'll make more tweaks.